### DISCUSSION TITLE
Muiltifile headline links

### DISCUSSION THREAD

**Discussion by pascalkuthe** [nvim-orgmode/orgmode#111] [Q&A] [answered]
Hi,

thanks for developing this great plugin :+1: 

I have a question regarding the orgmode gif shown in the README.
In the gif links to headings in files (the related work issues) are suggested by autocompletion and jumped to. However, this doesn't work for me. From looking at the sourcecode it also seems only the current file is searched for hyperlinks. How can I achieve similar behaviour as shown in the gif?

**Comment by levouh** [comment]
Per what I can see from the completion [here](https://github.com/kristijanhusak/orgmode.nvim/blob/7592652e72cb1a69bd147e118fc05d373f09045b/lua/orgmode/org/hyperlinks.lua#L16), it looks like it only searches the current file for links.

Worth opening up an issue to track it as a new feature @kristijanhusak?

**Reply by kristijanhusak** [reply]
@DSPOM2 @levouh Initially it supported that, but once I added support for exporting, I noticed that emacs orgmode doesn't support cross file linking, and fails to export. Here's the commit that deleted the support https://github.com/kristijanhusak/orgmode.nvim/commit/9c1285b1cf048c3b40a2d8f5be6ff598b0a619bd. I wasn't aware of this limitation once I made the Gifs.

**Reply by levouh** [reply]
Gotcha. I'll have to actually get _Emacs_ org-mode setup to be able to do testing like this, but the best option for now seems to just be a `[[file:...]]` link to do this. I know people like things like `org-roam` to link between files, so I imagine there _has_ to be a way to create cross-file links to specific sections. Seems like a relatively critical feature to have, but I'll see if I can play with org-mode a bit more to see how its done there.

**Reply by levouh** [reply]
A super quick search yields that:

```
[[file:work.org::*|]]
```

(cursor placed at `|`) ~~actually does what is wanted here.~~

---

Edit: I had two headers with the same title, so it only _looked_ to be working when it actually isn't.

**Reply by levouh** [reply]
Thoughts on adding "org files" to completion @kristijanhusak? This would then support a sort of "chained" completion style where you first select the file, then a header within that file to link to or other text. Looks like _Emacs_ supports external linking via something like:

- Words text search in `.org` file:
```
file:projects.org::some|
```

- Task title heading search in `.org` file
```
file:projects.org::*|
```

Not entirely sure what the best way to go about knowing "the file before the cursor" would be. I'm only familiar with adding sources to `nvim-cmp`, where the text before the cursor could be searched for `file:<path>::`, and then provide completion based off of "other file" versus "current file".

**Reply by kristijanhusak** [reply]
@levouh is this also supported in the emacs orgmode? If yes, and if it doesn't break the export, yeah, why not.

**Reply by levouh** [reply]
Parsing the file is relatively trivial, just a simple `string.match(line, 'file:(.-)::')`, but handling loading the file if it isn't already loaded becomes a bit more convoluted.

Got a super basic proof-of-concept working:

![image](https://user-images.githubusercontent.com/31262046/138604819-4325cbcd-f5d0-4013-9e9a-1065a02249b9.png)

**Reply by levouh** [reply]
@kristijanhusak as far as functionality support from _Emacs_ goes, see it being supported under the `file` section [here](https://orgmode.org/manual/External-Links.html). It notes you can do things like:

```
file:projects.org::some words
file:projects.org::*task title
file:projects.org::#custom-id
```

**Reply by kristijanhusak** [reply]
Thanks @levouh . Would you like to implement this yourself? You could set up a draft PR with your proof of concept and we could go from there.

**Reply by levouh** [reply]
Yup, poking at it right now.

Do you have an easy place to bounce ideas beyond Github (discussions/issues)? Perhaps Matrix or similar?

There are a number of ways to handle "hot loading" files (say the user is trying to link a file that isn't loaded yet) and the way you have it implemented right now is callback-style (in a context where a result need be returned).

It's pretty easy to just load the file on one call, and provide completion results in the next typed character but it doesn't make for a super good user experience. Its relatively easy to do an await/promise style handoff, but we end up crossing the Lua/C boundary so that doesn't work super well either.

Wanted to discuss w/ you @kristijanhusak before going too far down one path or considering something like Plenary's async module.

**Reply by kristijanhusak** [reply]
I created Matrix today, but I'm not really used to using it yet.

If the file you are trying to link is in the `org_agenda_files` path, you don't have to worry about it. Once orgmode is loaded (agenda, capture, or any org file opened) it loads all of files from there automatically. I'm not sure if linking to something outside of it (org files, not other files) should even work. If you can, give it a test with Emacs orgmode. I'm really trying to strictly follow what's happening there.

I'm ok with having await/promise style, but I don't want to create a dependency on Plenary.

I suggest to set up a PR with the assumption that file will always be loaded into Orgmode, and we can continue discussion from there.

**Reply by pascalkuthe** [reply]
Sry for not replying earlier, it seems Github doesn't send me any notifications for this thread.
Thanks, @levouh @kristijanhusak  for the quick answer. I was planning to help with the implementation myself but it seems you were faster. I will test the PR in my setup to see if I can find any bugs. May I suggest updating the GIF in the README in  #112 asweell?

**Comment by levouh** [comment, marked as answer]
See [here](https://github.com/kristijanhusak/orgmode.nvim/pull/112).
